code_tools/2 type_checkers/index.md

```diff
@@ -0,0 +1,570 @@
+
+
+    isinstance(a, int)
+
+</details>
+
+<details markdown="1"><summary  markdown="span">Type Checking</summary>
+
+Type information can be used for different things, such as optimizing programs, ensuring enough memory is available, but perhaps most important to us programmers: type checking. Ensuring that the program is free of any type errors. 
+
+Type checking can be done both dynamically and statically. But, dynamic type checking happens while running the program and will inevitably impact performance. That is why dynamic type checking is usually only done upon execution of a line of code, to ensure no unnecessary checks are done. That execution of a line of code might be very late, and if you are not testing properly, it might just be in the hands of the end user by the time that buggy line of code finally runs.
+
+In contrast static type checking does not need to worry about performance (as much). Afterall, this style of type checking happens in the developer's time (hence the "as much"), and not the end user's time. This enables a static type checker to do more complex type of checks and give better hints as to what is going wrong. However, to perform static type checking the information on types needs to be available before running the code. That often means that you as a programmer need to add this information. That is both a blessing and a curse. You will have to write more code, but the added information will make it easier for your co-workers or your future self to understand.
+
+Static type checking is often preferred, and so much so that languages such as JavaScript (in the form of TypeScript) and Python have started to adopt type information to enable static type checking. 
+</details>
+
+
+## Type hints in Python
+
+<details markdown="1"><summary  markdown="span">Type hints</summary>
+A type hint in the simplest form looks like this:
+
+
+    foo: int
+
+
+All this says is, the type of the variable `foo` should be an integer. Notice how there is no initial value here. This line of code does not create a variable `foo`, all it does is add a hint that `foo`, once it exists, should be an integer. That means this will raise a `NameError`:
+
+
+    $ python
+    >>> foo: int
+    >>> foo
+    NameError: name 'foo' is not defined
+
+
+It is possible to combine type hints and initialization on the same line, like so:
+
+
+    foo: int = 3
+
+
+That looks somewhat redundant, doesn't it? How can the *literal* `3` be anything else than an integer? This is where type inference kicks in. Tools such as `mypy` will try to infer the types of variables from their use. It is quite safe to assume type inference is possible here, so probably best to just write:
+
+
+    foo = 3
```

Comment:
Misschien met stukje juist beginnen met dit voorbeeld? Dan kun je nog even kort uitleggen: hoe weet Python dat het een int is? Omdat er een int literal wordt gebruikt. (ze hebben nauwelijks het woord 'literal' gehad dus misschien nog sterker definieren ook)